animation/abilities/Ability.java

```diff
@@ -0,0 +1,179 @@
+package animation.abilities;
+
+import SquarePG.SquarePG;
+import animation.Animation;
+import animation.effects.MapEffect;
+import animation.effects.Projectile;
+import characterEntities.Entity;
+import screens.Drawable;
+
+import javax.swing.*;
+import java.awt.*;
+import java.util.ArrayList;
+
+public abstract class Ability implements Drawable {
+	public enum AbilityState {
+		INITIALIZING,
+		CAN_DAMAGE,
+		HAS_EXPIRED,
+		IS_DONE
+	}
+
+	public static final String FILEPATH_ABILITY = "abilities/";
+
+	protected Animation initializeAnimation = null;
+	protected Animation canDamageAnimation = null;
+	protected Animation expirationAnimation = null;
+	protected Entity entity = null;
+	private Entity.EntityAbility entityAbility;
+	protected AbilityState state = AbilityState.INITIALIZING;
+	protected int cooldownTotal, cooldownCounter;
+
+	protected ArrayList<Projectile> projectiles = null;
+
+	public Ability(Entity entity, double cooldownInSeconds, Entity.EntityAbility entityAbility) {
+		this.cooldownCounter = 0;
+		this.entity = entity;
+		this.cooldownTotal = (int)Math.round(cooldownInSeconds * SquarePG.FPS);
+		this.entityAbility = entityAbility;
+	}
+
+	public void update() {
+		//Updating the state of the ability
+		if (state == AbilityState.INITIALIZING && (initializeAnimation == null || initializeAnimation.isDone())) {
+			resetCooldown();
+			if (canDamageAnimation != null) {
+				state = AbilityState.CAN_DAMAGE;
+			} else if (expirationAnimation != null) {
+				state = AbilityState.HAS_EXPIRED;
+			} else {
+				state = AbilityState.IS_DONE;
+			}
+		} else if (state == AbilityState.CAN_DAMAGE && (canDamageAnimation == null || canDamageAnimation.isDone())) {
+			if (expirationAnimation != null) {
+				state = AbilityState.HAS_EXPIRED;
+			} else {
+				state = AbilityState.IS_DONE;
+			}
+		} else if (state == AbilityState.HAS_EXPIRED && (expirationAnimation == null || expirationAnimation.isDone())) {
+			state = AbilityState.IS_DONE;
+		}
+
+		Animation currentAnimation = null;
+		switch(state) {
+			case INITIALIZING:
+				currentAnimation = initializeAnimation;
+				break;
+			case CAN_DAMAGE:
+				currentAnimation = canDamageAnimation;
+				break;
+			case HAS_EXPIRED:
+				currentAnimation = expirationAnimation;
+				break;
+			case IS_DONE:
+				break;
+			default:
+				break;
+		}
+
+		if (currentAnimation != null) {
+			currentAnimation.update();
+			int additionalX = entity.getFacingEast()? 0: entity.getImageIcon().getIconWidth()-2*currentAnimation.getOffsetX()-currentAnimation.getSize().width;
```

Comment:
why don't we just double the offset value instead of multiplying it by 2 every time